File: app/javascript/channels/user_channel.js
import { createUserConsumer, username } from "channels/consumer"
import {appendRoom, appendUser} from "../lobby";



const types = {
  currentUsers: "ONLINE_USERS",
  currentRooms: "ACTIVE_ROOMS",
  joinRequest: "JOIN_REQUEST",
  viewRequest: "VIEW_REQUEST"
}

export function requestToJoin(guest, host, room) {
  return  userChannel.send({
    type: types.joinRequest,
    data: {
      guest: guest,
      host: host,
      room: room
    }
  })
}

export function requestToView(guest, host, room) {
  return  userChannel.send({
    type: types.viewRequest,
    data: {
      guest: guest,
      host: host,
      room: room
    }
  })
}

export const userChannel = createUserConsumer.subscriptions.create({channel: "UserChannel", user: username}, {
  connected() {
    console.log("user connected")
  },

  disconnected() {

  },

  received(response) {
    console.log(response)
    switch (response.type) {
      case types.currentRooms:
        return  response.data.forEach(room => appendRoom(room))
      case types.currentUsers:
        return  response.data.forEach(user => appendUser(user))
      case types.joinRequest:
        return Snackbar.show({pos: 'bottom-right',
          duration: 5000, text: `${response.data.guest} requested to join room ${response.data.room}`})
      case types.viewRequest:
        return Snackbar.show({pos: 'bottom-right',
          duration: 5000, text: `${response.data.guest} requested to view room ${response.data.room}`})
    }
  }
});

File: app/channels/user_channel.rb
class UserChannel < ApplicationCable::Channel
  require 'chess_in_mem_cache'

  @@cache = ChessInMemCache.instance

  @@message_types = {
    current_rooms: "ACTIVE_ROOMS",
    current_users: "ONLINE_USERS",
    join_request: "JOIN_REQUEST",
    view_request: "VIEW_REQUEST"
  }

  def subscribed
    stream_from "user_channel_#{params[:user]}"
    send_subed_users
    send_active_rooms
  end

  def receive(request)
    case request["type"]
      when @@message_types[:join_request]
        then handle_join_request(request["data"]["guest"], request["data"]["host"], request["data"]["room"])
      when @@message_types[:view_request]
        then handle_view_request(request["data"]["guest"], request["data"]["host"], request["data"]["room"])
    end
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def send_subed_users
    cache = @@cache.all
    users = cache.is_a?(Array) ? cache : cache.to_a
    ActionCable.server.broadcast "user_channel_#{params[:user]}", {type: @@message_types[:current_users], data: users}
  end

  def send_active_rooms
    rooms = Room.all
    ActionCable.server.broadcast "user_channel_#{params[:user]}", {type: @@message_types[:current_rooms], data: rooms}
  end

  def handle_join_request(guest, host, room)
    ActionCable.server.broadcast "user_channel_#{host}", { type: @@message_types[:join_request], data: { guest: guest, room: room }}
  end

  def handle_view_request(guest, host, room)
    ActionCable.server.broadcast "user_channel_#{host}", { type: @@message_types[:view_request], data: { guest: guest, room: room }}
  end
end
